topdown/copypropagation/copypropagation.go

```diff
@@ -289,34 +295,44 @@ func (p *CopyPropagator) updateBindingsEqAsymmetric(a, b *ast.Term) (ast.Var, as
 }
 
 type plugContext struct {
-	bindings map[ast.Var]*binding
-	uf       *unionFind
-	headvars ast.VarSet
-	negated  bool
+	removedEqs *ast.ValueMap
+	uf         *unionFind
+	headvars   ast.VarSet
+	negated    bool
 }
 
 type binding struct {
-	k ast.Var
+	k ast.Value
 	v ast.Value
 }
 
-func newbinding(k ast.Var, v ast.Value) *binding {
-	return &binding{k: k, v: v}
+func safelyContainedIn(value ast.Value, query ast.Body) bool {
```

Comment:
Hrmmm. I'm not sure about this function. I think we're compensating for removal of the negation/groundness check from the plugBindings function? What I'm unsure about is what this function is intended to do. Safety ensures that all variables appear as an output in at least one non-negated expression. As is, the logic here does not quite cover that. Perhaps we should just expose the safety check or the output variable analysis from the ast package so that it can be reused here. I've got some pending changes to that code that refactor it and include tests. We could merge that into master first (along with the union-find refactoring in this PR--which is great btw) and then fix this problem. If we do that, I could imagine just updating the plugBindings logic alone.